### PR TITLE
Convert xcm to lowered form

### PR DESCRIPTION
**Pull request by alistair-singh**
Cumulus Companion: https://github.com/Snowfork/cumulus/pull/17

There is a bridge config aspect to the bridge-transfer pallet. You can add a target_location_fee which is the fee charged for execution on the target chain.

https://github.com/paritytech/cumulus/pull/2013/files#diff-e33e5afc30ebd89df96d4630222efd1c686a38185d23c958d94d5f6fbebdf6feR679-R682

Here is where it is configured for rococo-wococo
https://github.com/paritytech/cumulus/blob/bridge-hub-rococo-wococo/scripts/bridges_rococo_wococo.sh#L192-L255
https://github.com/paritytech/cumulus/blob/bridge-hub-rococo-wococo/scripts/bridges_rococo_wococo.sh#L574-L583

In the case of rococo-wococo the target fee is 50000000000 Wococo. For our bridge, we would need to set target_location_fee to None because we do not have a way to specify remote fees on our bridging protocol.
As for the fees between Statemine and BridgeHub, It seems they are using UnpaidRemoteExporter. So fees are not specified here either.

https://github.com/paritytech/cumulus/blob/bko-transfer-asset-via-bridge/parachains/runtimes/assets/statemine/src/xcm_config.rs#L506
https://github.com/paritytech/cumulus/blob/bko-transfer-asset-via-bridge/parachains/runtimes/assets/statemine/src/lib.rs#L708

So our bridge config would be:

https://github.com/paritytech/cumulus/blob/bko-transfer-asset-via-bridge/parachains/pallets/bridge-transfer/src/lib.rs#L72-L87
```
{
	bridge_location: "../Parachain(1013)"                                      // Bridgehubs para id
	bridge_location_fee: None                                                  // Because we are using UnpaidRemoteExporter which asserts there is no fee.
	allowed_target_location: "../../GlobalConsensus(Ethereum{ chain_id: 1})"
	max_target_location_fee: None                                              // Because we have no way of sending this information yet.
}
```

Pseudo code for XCM packet building in the bridge-transfer pallet:

```rs
if target_location_fee {
	WithdrawAsset(bridge_config.target_location_fee.clone().into()),
	BuyExecution { fees: target_location_fee, weight_limit: Unlimited },
} else {
	UnpaidExecution { check_origin: None, weight_limit: Unlimited },
}
ReserveAssetDeposited(reserved_assets.clone().into())
ClearOrigin
DepositAsset { assets: MultiAssetFilter::from(MultiAssets::from(reserved_assets)), beneficiary: remote_destination }
```
`remote_destination` here is an ethereum account: `../../GlobalConsensus(Ethereum{ chain_id: 1})/AccountId20(0x1234567…)`

The HaulBlobExporter will prepend the following XCM before sending it across the bridge:
```
UniversalOrigin(GlobalConsensus(local_net)) - local_net will be Polkadot, Kusama, etc…
DescendOrigin(local_sub) - local_sub will be the para_id of statemine.
```

We are rewriting HaulBlobExporter so we do not need to match these. These are passed in parameters and we only need to validate them.

So the final XCM packet that the code will match to an Unlock is:
```
UnpaidExecution { check_origin: None, weight_limit: Unlimited }
ReserveAssetDeposited(reserved_assets.clone().into())
ClearOrigin
DepositAsset { assets: MultiAssetFilter::from(MultiAssets::from(reserved_assets)), beneficiary: remote_destination }
```
The code will have to check if all assets are native to ethereum because we only support Unlock on NativeTokens in the bridging protocol.

### CODE REVIEW

File: parachain/primitives/router/src/export.rs [hunk 1]
```diff
@@ -0,0 +1,1121 @@
...
+use ethabi::{self, Token};
+use frame_support::{ensure, log, traits::Get};
+use snowbridge_core::{ParaId, SubmitMessage};
+use sp_core::{RuntimeDebug, H160};
+use sp_std::{marker::PhantomData, prelude::*};
+use xcm::v3::prelude::*;
+use xcm_executor::traits::ExportXcm;
+
+#[derive(Encode, Decode)]
+struct BridgeMessage(ParaId, u16, Vec<u8>);
```
Comment (alistair-singh) [inline parachain/primitives/router/src/export.rs:13]: We need to encode the parachain id, handler id and the payload. This type lives to hold these temporarily between `ExportXcm` traits `validate` and `deliver` functions.

File: parachain/primitives/router/src/export.rs [hunk 2]
```diff
@@ -0,0 +1,1121 @@
...
+
+#[derive(PartialEq, RuntimeDebug)]
+enum NativeTokens {
+	Unlock { asset: H160, destination: H160, amount: u128 },
+}
+
+#[derive(PartialEq, RuntimeDebug)]
+enum OutboundPayload {
+	NativeTokens(NativeTokens),
+}
```
Comment (alistair-singh) [inline parachain/primitives/router/src/export.rs:119]: Every type of message we send out will be described in this enum. 

Conversion will flow like:
```
Xcm<()> -> OutboundPayload -> (payload, handler)
```

File: parachain/primitives/router/src/export.rs [hunk 3]
```diff
@@ -0,0 +1,1121 @@
...
+	ZeroAssetTransfer,
+	BeneficiaryResolutionFailed,
+	AssetResolutionFailed,
+}
+
+struct XcmConverter<'a, Call> {
+	iter: Iter<'a, Instruction<Call>>,
+	bridged_location: &'a NetworkId,
+}
+impl<'a, Call> XcmConverter<'a, Call> {
```
Comment (alistair-singh) [inline parachain/primitives/router/src/export.rs:162]: Converting an xcm message to an outbound payload is handled here.
```
Xcm<()> -> OutboundPayload
```

File: parachain/primitives/router/src/export.rs [hunk 4]
```diff
@@ -0,0 +1,1121 @@
...
+	Unlock { asset: H160, destination: H160, amount: u128 },
+}
+
+#[derive(PartialEq, RuntimeDebug)]
+enum OutboundPayload {
+	NativeTokens(NativeTokens),
+}
+
+impl OutboundPayload {
+	fn abi_encode(&self) -> (Vec<u8>, u16) {
```
Comment (alistair-singh) [inline parachain/primitives/router/src/export.rs:122]: Resolving handlers and encoding payloads is done here:
```
OutboundPayload -> (payload, handler)
```

**Comment by vgeddes** [comment]
Looks very good! Will review more deeply a bit later this week.

File: core/packages/contracts/src/NativeTokens.sol [hunk 5]
```diff
@@ -14,7 +15,7 @@ import { ParaID } from "./Types.sol";
 /// @title Native Tokens
 /// @dev Manages locking, unlocking ERC20 tokens in the vault. Initializes ethereum native
 /// tokens on the substrate side via create.
-contract NativeTokens is AccessControl {
+contract NativeTokens is AccessControl, IRecipient {
```
Comment (vgeddes) [inline core/packages/contracts/src/NativeTokens.sol:18]: nice catch!

File: parachain/pallets/outbound-queue/src/lib.rs [hunk 6]
```diff
@@ -161,9 +161,9 @@ pub mod pallet {
 		}
 	}
 
-	impl<T: Config> Pallet<T> {
+	impl<T: Config> SubmitMessage for Pallet<T> {
 		/// Submit message on the outbound channel
-		pub fn submit(origin: &ParaId, handler: u16, payload: &[u8]) -> DispatchResult {
+		fn submit(origin: &ParaId, handler: u16, payload: &[u8]) -> DispatchResult {
```
Comment (vgeddes) [inline parachain/pallets/outbound-queue/src/lib.rs:166]: `ParaId` is really just struct with a single `u32` field. Surely it implements `Copy`, in which case there is no need to pass by reference

File: parachain/primitives/core/src/lib.rs [hunk 7]
```diff
@@ -26,3 +27,7 @@ pub trait Verifier {
 		descendants_until_final: u8,
 	) -> Result<(), &'static str>;
 }
+
+pub trait SubmitMessage {
```
Comment (vgeddes) [inline parachain/primitives/core/src/lib.rs:31]: this trait name is a bit too simple. For example it doesn't describe where the message is being submitted.

```suggestion
pub trait OutboundQueue {
```

File: parachain/primitives/router/src/export.rs [hunk 8]
```diff
@@ -0,0 +1,1121 @@
...
+
+	fn deliver((blob, hash): (Vec<u8>, XcmHash)) -> Result<XcmHash, SendError> {
+		let mut blob = blob.clone();
+		let mut input: &[u8] = blob.as_mut();
+		let BridgeMessage(source_id, handler, payload) = BridgeMessage::decode(&mut input)
+			.map_err(|err| {
+				log::trace!(target: "ethereum_blob_exporter", "undeliverable due to decoding error '{err:?}'.");
+				SendError::NotApplicable
+			})?;
+		Submitter::submit(&source_id, handler, payload.as_ref()).map_err(|err| {
```
Comment (vgeddes) [inline parachain/primitives/router/src/export.rs:102]:
```suggestion
		OutboundQueue::submit(origin, handler, payload.as_ref()).map_err(|err| {
```

**Review by vgeddes — COMMENTED** [review]
Nice work Alistair! I appreciate the extensive unit tests.

Added a few minor comments mostly about naming.

**Comment by yrong** [comment]
> So the final XCM packet that the code will match to an Unlock is:

```
UnpaidExecution { check_origin: None, weight_limit: Unlimited }
ReserveAssetDeposited(reserved_assets.clone().into())
ClearOrigin
DepositAsset { assets: MultiAssetFilter::from(MultiAssets::from(reserved_assets)), beneficiary: remote_destination }
```

AFAIK `ReserveAssetDeposited` is only used for transfer self reserve asset to another chain(e.g. DOT from statemint through bridgehub to Ethereum), for `unlock` here is actually withdraw/burn the asset from ethereum so xcm should be something like:

```
WithdrawAsset(eth_assets),
InitiateReserveWithdraw {
	assets: All.into(),
	reserver: bridge_hub.into(),
	xcm: Xcm(vec![
		BuyExecution(),
		DepositAsset(recipient(H160)),
	]),
},
``` 

`WithdrawAsset` will just burn the minted asset in statemint and `InitiateReserveWithdraw` will triger a xcm call to bridgehub,  and when it comes to bridgehub, only the internal xcm(i.e. DepositAsset) is left.

so I'm not sure the convert logic [here](https://github.com/Snowfork/snowbridge/blob/1ddc268808a458fd81d5fa0d6e30726f21782057/parachain/primitives/router/src/export.rs#L201) is correct.

File: parachain/pallets/outbound-queue/src/lib.rs (hunk 6 above)
Comment (alistair-singh) [reply parachain/pallets/outbound-queue/src/lib.rs]: addressed in [da6a3e0](https://github.com/Snowfork/snowbridge/pull/805/commits/da6a3e0ed03ef96c989bc8a54d1383afa2523286)

File: parachain/primitives/core/src/lib.rs (hunk 7 above)
Comment (alistair-singh) [reply parachain/primitives/core/src/lib.rs]: addressed in [1a59f22](https://github.com/Snowfork/snowbridge/pull/805/commits/1a59f22c260a53df423de9a590b31b55ab1c9627)

File: parachain/primitives/router/src/export.rs [hunk 9]
```diff
@@ -0,0 +1,1121 @@
...
+use xcm::v3::prelude::*;
+use xcm_executor::traits::ExportXcm;
+
+#[derive(Encode, Decode)]
+struct BridgeMessage(ParaId, u16, Vec<u8>);
+
+pub struct ToBridgeEthereumBlobExporter<RelayNetwork, BridgedNetwork, Submitter>(
+	PhantomData<(RelayNetwork, BridgedNetwork, Submitter)>,
+);
+impl<RelayNetwork: Get<NetworkId>, BridgedNetwork: Get<NetworkId>, Submitter: SubmitMessage>
```
[the remark this answers is not in the document]
Comment (alistair-singh) [reply parachain/primitives/router/src/export.rs]: addressed in [192a9c7](https://github.com/Snowfork/snowbridge/pull/805/commits/192a9c7b15ea7dadbd08822c4445ab71f329e5a6)

File: parachain/primitives/router/src/export.rs [hunk 10]
```diff
@@ -0,0 +1,1121 @@
...
+use snowbridge_core::{ParaId, SubmitMessage};
+use sp_core::{RuntimeDebug, H160};
+use sp_std::{marker::PhantomData, prelude::*};
+use xcm::v3::prelude::*;
+use xcm_executor::traits::ExportXcm;
+
+#[derive(Encode, Decode)]
+struct BridgeMessage(ParaId, u16, Vec<u8>);
+
+pub struct ToBridgeEthereumBlobExporter<RelayNetwork, BridgedNetwork, Submitter>(
```
[the remark this answers is not in the document]
Comment (alistair-singh) [reply parachain/primitives/router/src/export.rs]: addressed in https://github.com/Snowfork/snowbridge/pull/805/commits/2a72daf9c1d8bec9316b4975d2f9efc71f0a35a0

File: parachain/primitives/router/src/export.rs [hunk 11]
```diff
@@ -0,0 +1,1121 @@
...
+		log::info!(target: "ethereum_blob_exporter", "message validated {hash:#?}.");
+
+		// TODO: Fees if any currently returning empty multi assets as cost
+		Ok(((blob, hash), MultiAssets::default()))
+	}
+
+	fn deliver((blob, hash): (Vec<u8>, XcmHash)) -> Result<XcmHash, SendError> {
+		let mut blob = blob.clone();
+		let mut input: &[u8] = blob.as_mut();
+		let BridgeMessage(source_id, handler, payload) = BridgeMessage::decode(&mut input)
```
[the remark this answers is not in the document]
Comment (alistair-singh) [reply parachain/primitives/router/src/export.rs]: addressed in [d50e04d](https://github.com/Snowfork/snowbridge/pull/805/commits/d50e04d1ac6a16455c10739778636e88600d6173)

**Comment by alistair-singh** [comment]
@yrong Your example is correct in a more general xcm way.  But bridges are configured, and transfers are initiated from `statemine` with the `bridge-transfer` [pallet](https://github.com/paritytech/cumulus/blob/43ee62707d80b46a954a5742117b258b6996f3a2/parachains/pallets/bridge-transfer/src/lib.rs#LL304C10-L304C35) using `transfer_asset_via_bridge` extrinsic. 

This forum post has an example for substrate to substrate chains. Wococo to Rococo. It shows the xcm that `transfer_asset_via_bridge` produces and sends to bridgehub. This is the message the `EthereumExporter` should match for unlocks.
https://forum.polkadot.network/t/bridge-hub-questions/2693/5
If the asset comes from ethereum i.e. `GlobalConsensus(Ethereum { chain_id: 1 })` then the code must route to NativeTokens smart contract handler. If the asset comes from substrate i.e. `GlobalConsensus(Rococo)` then the code must route to another smart contract handler which knows how to handle substrate assets.

The locking(reserving) of assets is currently implemented in the `bridge-transfer`.
https://github.com/paritytech/cumulus/blob/43ee62707d80b46a954a5742117b258b6996f3a2/parachains/pallets/bridge-transfer/src/lib.rs#L630-L637

**Comment by yrong** [comment]
@alistair-singh Gotcha, and saw your comment here https://github.com/paritytech/cumulus/pull/2013#discussion_r1171077178
seems still some revamp required in that PR.

**Review by vgeddes — APPROVED** [review]

**Review by yrong — APPROVED** [review]
LGTM!

File: core/packages/test/config/launch-config.toml [hunk 12]
```diff
@@ -51,7 +51,7 @@ cumulus_based = true
...
+    args = ["--enable-offchain-indexing=true", "--pruning=archive", "--force-authoring", "-lparachain=trace,cumulus-collator=trace,aura=trace,xcm=trace,ethereum-beacon-client=trace,runtime::bridge-hub=trace,runtime::bridge=trace,runtime::bridge-dispatch=trace,bridge=trace,runtime::bridge-messages=trace,xcm::weight=trace,xcm::filter_asset_location=trace,xcm::send_xcm=trace,xcm::barriers=trace,xcm::barrier=trace,xcm::execute_xcm=trace,xcm::contains=trace,xcm::execute_xcm_in_credit,xcm::process_instruction=trace,xcm::currency_adapter=trace,xcm::origin_conversion=trace,xcm::fungibles_adapter=trace,xcm::process=trace,xcm::execute=trace,xcm::should_execute=trace,runtime=debug,ethereum_blob_exporter=trace"]
```
Comment (yrong) [inline core/packages/test/config/launch-config.toml:55]: Nit: seems a bit too long, will `xcm=trace`  just be fine?

**Pull request merged**